### PR TITLE
Log out button

### PR DESCRIPTION
**Pull request by Jipje**
Work in progress.

Issues:

`<lancie-logout-button hidden$="[[!user]]"></lancie-logout-button>`  is not sufficient to hide the logout button when a user is not logged in.

Also we need to find a way to refresh the "cache" or local storage after logging out. Logging out brings you to the homepage but returning to the My Area page still shows the users information. Using ```window.location.replace("/");``` instead of a `navigate` event seemed to work...

### CODE REVIEW

**Comment by Fastjur** [comment]
So. The issue here is that the `iron-lazy-pages` loads the pages when clicked, but they are not removed after logging out. A `window.location.replace("/")` call works because the full page is then reloaded (and the `iron-lazy-pages` need to be reloaded again (causing the `login-check` to fail and thus showing a login).

I'm tagging @TimvdLippe  as he's the author of `iron-lazy-pages`. I know there is a way to force the inputs etc. to reload always, but I was unable to find a way to force them to be 'unloaded' or something similar.

File: src/lancie-storage/lancie-auth-storage.html [hunk 1]
```diff
@@ -58,6 +58,20 @@
...
+        this.$.form.clearToken();
+      },
+      // TODO include these functions in their respective repositories.
+      // Feel free to use the following code for testing purposes
+      //
+      // clearToken: function() {
+      //  this.auth = {};
+      //  this.$.ajax.injectToken('');
+      //  this.$.form.injectToken('');
+      // },
```
Comment (Fastjur) [inline src/lancie-storage/lancie-auth-storage.html]: This PR works as intended after we fix these issues

File: index.html [hunk 2]
```diff
@@ -186,6 +186,7 @@
               <a data-route="tickets" href="/tickets">Tickets</a>
               <a data-route="my-area" href="/my-area">My Area</a>
               <a data-route="seatmap" href="/seatmap">Seatmap</a>
+              <lancie-logout-button hidden$="[[!user]]"></lancie-logout-button>
```
Comment (TimvdLippe) [inline index.html]: Is the CSS style not applied? In that case, you have to do `[hidden] { display: none !important; }`

File: src/lancie-login/lancie-logout-button.html [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      min-height: 48px;
+      height: inherit;
+      display: flex;
+      align-items: center;*/
+    }
+  </style>
+
+  <lancie-ajax id="ajaxLogout" refurl="logout" method="POST" on-lancie-ajax="onLogoutResponse"></lancie-ajax>
+
+  <a href="#" on-tap="tryLogout">Logout</a>
```
Comment (TimvdLippe) [inline src/lancie-login/lancie-logout-button.html]: Shouldn't this be a `<button>` instead?

File: src/lancie-login/lancie-logout-button.html [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+</template>
+<script>
+  (function () {
+  'use strict';
+
+  const authStorage = document.querySelector('lancie-auth-storage');
```
Comment (TimvdLippe) [inline src/lancie-login/lancie-logout-button.html]: This is brittle, as we do not know whether this remains. In general: do not query outside from your element, only going down.

File: src/lancie-login/lancie-logout-button.html [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+  class LancieLogoutButton extends Polymer.Element {
+    static get is() {
+      return 'lancie-logout-button';
+    }
+
+    static get properties() {
```
Comment (TimvdLippe) [inline src/lancie-login/lancie-logout-button.html]: I think you can omit this if this element has no properties.

**Comment by martijnjanssen** [comment]
The easiest solution to all of this is removing all the functionality we added and just calling the logout endpoint and refreshing the page. Doing it this way, the page takes care of all the expired tokens and not giving access to users not logged in. Way easier, sorry for only realizing this now.

**Comment by Fastjur** [comment]
Yes that would be easiest. However I recall that most adblock plugins prevent a page from refreshing, and some browsers too. I think we would end up not having this work in 10/30% (educated guess) of the users.

**Comment by martijnjanssen** [comment]
But as discussed, there isn't any perfect solution to circumvent this, we we're just going with the solution I proposed right?

**Comment by Fastjur** [comment]
Yes after the discussion in Slack I think that is the best approach!

File: index.html (hunk 2 above)
Comment (elarb) [reply index.html]: This actually won't work at all since it won't bind to anything.

**Comment by elarb** [comment]
I moved the button to My Area:

logout-button

**Comment by elarb** [comment]
What is the state of this PR? Do we want to look more into where to place the button?

**Comment by Fastjur** [comment]
Wouldn't we want to move this to the header? In the top right?

It should be possible to hide it when not logged in right.

**Comment by elarb** [comment]
@Fastjur Yes, I think we should do that. I took a look at it earlier today but something else caught my attention, but will look at it again later

**Comment by Fastjur** [comment]
Alright thank you!

Op ma 12 feb. 2018 om 18:16 schreef e_lar <notifications@github.com>:

**Comment by TimvdLippe** [comment]
Closing this, as rebasing becomes very nasty against current master. Let's retry this with a fresh PR.

**Pull request closed without merging**